### PR TITLE
Improve precision of datetime index using TIMESTEP

### PR DESCRIPTION
**Pull request by asnyv**
Enhancement to tackle the issue described in #391

### CODE REVIEW

**Review by sigurdp — APPROVED** [review]
Good to go

**Review by sigurdp — APPROVED** [review]

File: ecl2df/summary.py [hunk 1]
```diff
@@ -435,6 +432,49 @@ def df(
         )
         logger.warning("Duplicates: %s", list(dframe.columns[dupes]))
         dframe = dframe.loc[:, ~dframe.columns.duplicated()]
+
+    index_duplicates = dframe.index.duplicated(keep="first")
+    if any(index_duplicates):
```
Comment (berland) [inline ecl2df/summary.py:437]: Perhaps this should be separated out to a testable auxiliary function?

Comment (asnyv) [reply ecl2df/summary.py:447]: Done, also wrote some tests (including some additional testdata), though they run on the full `df()` function

File: ecl2df/summary.py [hunk 2]
```diff
@@ -435,6 +431,55 @@ def df(
...
+
+    if datetime is True:
+        if dframe.index.dtype == "object":
+            dframe.index = pd.to_datetime(dframe.index)
+
+    return dframe
+
+
+def _ensure_unique_datetime_index(dframe: pd.DataFrame) -> pd.DataFrame:
+    index_duplicates = dframe.index.duplicated(keep="first")
```
Comment (berland) [inline ecl2df/summary.py:445]: docstring?

File: ecl2df/summary.py [hunk 3]
```diff
@@ -435,6 +431,55 @@ def df(
...
+        if "TIMESTEP" in dframe:
+            logger.info(
+                "Dataframe of smry data contained duplicate timestamps due to limited",
+                " output resolution. Vector TIMESTEP exists, utilizing it to create "
+                "discrete timestamps",
+                f"Original duplicates were:{index_duplicate_log_string}",
+            )
+            index_as_list = dframe.index.to_list()
+
+            if dframe.attrs["meta"]["TIMESTEP"]["unit"] == "DAYS":
```
Comment (berland) [inline ecl2df/summary.py:460]: I would guess you get a "KeyError" here if the unit is not available, so you will probably not go to the ValueError below. (?)

File: tests/test_summary.py [hunk 4]
```diff
@@ -33,6 +33,12 @@
 TESTDIR = Path(__file__).absolute().parent
 REEK = str(TESTDIR / "data/reek/eclipse/model/2_R001_REEK-0.DATA")
 EIGHTCELLS = str(TESTDIR / "data/eightcells/EIGHTCELLS.DATA")
+SHORT_STEP = str(TESTDIR / "data/timesteps/SHORT_STEP")
```
Comment (berland) [inline tests/test_summary.py:36]: Have these been made with EIGTHCELLS? Could these example UNSMRY be made reproducible?

File: ecl2df/summary.py (hunk 3 above)
Comment (asnyv) [reply ecl2df/summary.py:460]: The idea was to catch a scenario where unit exists, but not equal to DAYS or HOURS. It was not to catch a scenario where the unit key was missing. Think the existence of the unit key is pretty much ensured by running the smry_meta command and adding it to the dataframe (currently lines 418-422)

File: ecl2df/summary.py (hunk 2 above)
Comment (asnyv) [reply ecl2df/summary.py:445]: 😴  
but ok then 😉

File: tests/test_summary.py (hunk 4 above)
Comment (asnyv) [reply tests/test_summary.py:36]: Was made with the reek model, but doesn't really matter. Only thing was to have a schedule with forced steps, and as stripped UNSMRY files as possible. In the scenario without TIMESTEP: no vectors where requested in the SUMMARY section. In the scenarios with TIMESTEP: TIMESTEP was the only requested vector in the SUMMARY section. If I am to make it reproducible, I guess the EIGHTCELLS is better to make it light

File: ecl2df/summary.py (hunk 2 above)
Comment (asnyv) [reply ecl2df/summary.py:456]: done

**Review by berland — APPROVED** [review]

**Pull request merged**